### ISSUE TITLE
Support SSR applications

### ISSUE THREAD

**Issue by thure** [OfficeDev/microsoft-teams-library-js#602]
Not all Teams apps are SPA's - crucially, a project which supports server-side rendering will encounter build errors if this package is loaded since `window` is not defined.

This package should at least return some sort of default value instead of erroring when `window` is not defined.

**Comment by gablabelle** [comment]
We are facing the same issue. We are using next.js and in order to get pass these errors we have to make sure the library is only used client side.

@thure to get pass this, you can lazy import/require the library in a useEffect hook because it only runs client side or if you use next.js, use their `dynamic` component and disable ssr as per the [next.js documentation](https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr).

**Comment by Meghana-MSFT** [comment]
@thure Could you please share a sample or minimal code snippet for us to repro the issue?

**Comment by gablabelle** [comment]
> @thure Could you please share a sample or minimal code snippet for us to repro the issue?

Here you go https://stackblitz.com/edit/nextjs-avwbvb?file=pages%2Findex.js

Let stackblitz run the app install and display the working app on the right... Then uncomment the commented code below the TODO commnent and see that @microsoft/teams-js makes the app crash because it's trying to access `window` which doesn't exist when the code is running server-side (on nodejs).

The @microsoft/teams-js doesn't support SSR because of that.

**Comment by Meghana-MSFT** [comment]
microsoft/teams-js works in the context of Microsoft teams. Please try it out in teams context and it should work for you.

**Comment by gablabelle** [comment]
> microsoft/teams-js works in the context of Microsoft teams. Please try it out in teams context and it should work for you.

No @Meghana-MSFT you are wrong ... If the app embedded in the Teams iframe is using server-sider rendering it will break the same way even in the context of Teams.

This is because the library is trying to access the window object that doesn't exist server side. A check should be done prior to trying to access the window object.

**Comment by ryan11223344** [comment]
I have started developing a teams app for my group and I am running into the same problem, @gablabelle what did you do to bypass it for now until a patch can be made??

**Comment by gablabelle** [comment]
> I have started developing a teams app for my group and I am running into the same problem, @gablabelle what did you do to bypass it for now until a patch can be made??

@ryan11223344 You have to lazy load Teams client SDK within a `useEffect` hook because `useEffect` only runs client-side so you won't get the server-side error.

Something like the following in a React Functional component should work.
```
const [teamsClient, setTeamsClient] = useState();

useEffect(() => {
    const loadTeamsClient = async () => {
      const client = await import('@microsoft/teams-js');
      setTeamsClient(client);
    };
    loadTeamsClient();
}, []);
```

**Comment by Meghana-MSFT** [comment]
@thure 

Did you try any of the solutions and did it work out for you?

**Comment by shaohaolin** [comment]
+1 on this issue.
This also breaks on any SDK that has dependencies on microsoft/teams-js and then the SDK is used in any SSR applications.

**Comment by thure** [comment]
> @thure
> 
> Did you try any of the solutions and did it work out for you?

Sorry folks, I wasn't seeing notifications for this till now.

I'm not (currently) developing a Teams app, so I did not try any of these solutions, however @gablabelle's workaround would _work_, but obviously it does not solve the problem. Given that this package doesn't _need_ the `window` API to deliver strictly all meaningful functionality, it is an implementation flaw that this package can't run on the server.

If folks really need SSR with this package, it's all open source, someone should submit a PR.

**Comment by ChetanSharma-msft** [comment]
Hello Everyone,
Sorry for the late response.
Please let us know if you need any further help here or shall we close the issue?

**Comment by ChetanSharma-msft** [comment]
@gablabelle - Thanks for your response.
We will share your feedback/suggestion with engineering team & let you know the updates.

As an alternative, You can also raise your feedback here: https://feedbackportal.microsoft.com/feedback/forum/ad198462-1c1c-ec11-b6e7-0022481f8472

**Comment by gablabelle** [comment]
Quick message to prevent the issue from closing.

**Comment by nwojod-MSFT** [comment]
There is a backlog item for this issue, but no ETA for when it will be fixed. However, we are open if anyone from the community has a proposed fix and would like to submit a PR for it. Otherwise, we will update this thread when there is an ETA.

**Comment by GregfromKnowork** [comment]
Bump. Major blocker for our Teams app.

**Comment by DopamineDriven** [comment]
Huge blocker for my org. I have an embedded microsoft 3DVista Virtual tour "Tech refresh Store" with Azure AD auth wrapping it using Nextjs + NextAuth + an edge runtime middleware auth wrapper that works like a charm in the browser and when teh teams app is opened in the browser. The issue is that the vast majority of modern apps are isomorphic and use both server and client rendering. 

perhaps adding a conditionally invoked node runtime config to support SSR within teams would be a solution? Docker?

**Comment by nwojod-MSFT** [comment]
Hi @DopamineDriven sorry to hear this is a huge blocker for your org. We understand the importance of this feature, however, we are examining it with other features in our backlog and prioritizing accordingly.

**Comment by nwojod-MSFT** [comment]
We understand the importance of this feature and have added it to our backlog. The team will review it, the level of work required, and prioritize accordingly. Because of the volume of requests we receive, we are unable to promise when this feature will be implemented. However, we will provide an estimate on this thread as soon as we have a committed timeline. Thank you for your patience.

**Comment by webdes03** [comment]
```
const [teamsClient, setTeamsClient] = useState();
useEffect(() => {
    const loadTeamsClient = async () => {
        const client = await import("@microsoft/teams-js");
        setTeamsClient(client);
    };
    loadTeamsClient();
}, []);
```

I've got this working, but teamsClient is just an `any` type so it's a nightmare to work with. Is there a way to lazy import the types as well, or somehow reference the types externally and I just don't know how to do it? Maybe @gablabelle or @bsana1?

**Comment by TrevorJoelHarris** [comment]
Hi everyone! Sorry for the delay on helping out here. We'll actually be going in and making the fix @bsana1 mentioned above in the next couple of months. Part of the reason this change has taken a while to get to is that while making the fix @bsana1 mentions above is very straightforward, we also need to add some robust tests to make sure that no one inadvertently breaks SSR apps that use this library in the future.

**Comment by webdes03** [comment]
@TrevorJoelHarris is there any update on this fix or any idea where it sits in the prioritization? We've got numerous custom apps and products in the pipeline and want to move to an SSR model but are kind of stuck at the moment.

**Comment by TrevorJoelHarris** [comment]
> @TrevorJoelHarris is there any update on this fix or any idea where it sits in the prioritization? We've got numerous custom apps and products in the pipeline and want to move to an SSR model but are kind of stuck at the moment.

@webdes03 We are hoping to ship a fix for this in our next version (2.8) that will be coming out sometime next week. We are still validating that the fix works correctly, but assuming all goes well there we should be good to go.

**Comment by ryceg** [comment]
Is there any fix on this yet?

**Comment by TrevorJoelHarris** [comment]
Hi @ryceg, unfortunately getting validation in for this has taken longer than we were expecting for a bunch of boring reasons. We're just about finished, and are planning to ship a fix by the time the next official release comes out on 4/5. Sorry for the inconvenience!

**Comment by noahdarveau-MSFT** [comment]
Hello everyone, the fix for SSR has been merged into the repo and will be included in our next release of version 2.10.0 on 4/5.

**Comment by webdes03** [comment]
Thank you @noahdarveau-MSFT. I'm slow to follow up here, but can confirm teams-js working in nextjs with SSR. Thanks!

**Comment by guillewu** [comment]
@webdes03 could you share what do you mean that is working in nextjs with SSR? I found this example: https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/apps/ssr-test-app/pages/index.tsx#LL16C4-L16C4 and it seems like it's still initializing the library client side. If you have any tips, I'd appreciate it. Thanks!

**Comment by webdes03** [comment]
> @webdes03 could you share what do you mean that is working in nextjs with SSR? I found this example: https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/apps/ssr-test-app/pages/index.tsx#LL16C4-L16C4 and it seems like it's still initializing the library client side. If you have any tips, I'd appreciate it. Thanks!

I just used `npx create-next-app@latest` to create a Next solution from scratch, then created a page in `pages/tabs/app/config/index.tsx` to model a configuration page for a tab app. I'm using a fork of the useTeams hook at https://github.com/wictorwilen/msteams-react-base-component inside the page to fetch the teams context and return it to my page. If it seems like it's just stuck, make sure you're calling `notifySuccess` to tell the Teams SDK your page loaded.

**Comment by guillewu** [comment]
@thure @webdes03 thanks for your quick reply. I just moved some code to use SSR, ie:
```
import * as teamsjs from "@microsoft/teams-js";
const { authentication } = teamsjs;
const token = await authentication.getAuthToken();
```
And this no longer works as it's happening server side. Would you guys happen to know how to obtain a token now SSR? TIA

**Comment by webdes03** [comment]
I've always uses OBO in a server route to get tokens- in that pattern you'd get an identity token client side, and then hand off to an API route to convert that into a scoped access token.

**Comment by ricardasjak** [comment]
@webdes03 , @guillewu would you mind to share some example how to pass client side access token acquired by teamsjs to next-auth?

**Comment by webdes03** [comment]
I've actually never used next-auth. I always build my Teams apps to use SSO so by the time they get to my app they've already been let through the App Service's auth layer, and I only need the access tokens to call services like Graph. I just do the OBO flow server side instead of messing about with middleware. Does next-auth support OBO?
